File: cpp/delete-node-in-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode* next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// Note: we are assuming this never deletes the tail
class Solution {
public:
    // O(1)
    void deleteNode(ListNode* node) {
        // move the next value one node down
        node->val = node->next->val;

        ListNode* curr = node->next;

        // move the node's next pointer down one node
        node->next = node->next->next;

        delete curr;
        curr = nullptr;
    }
};

File: py/delete-node-in-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# O(1)
class Solution:
    def deleteNode(self, node: ListNode) -> None:
        # Given: The linked list will have at least two elements.
        # if node is None:
        #     return

        temp = node
        node.val = node.next.val
        node.next = node.next.next
        del temp
